src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -3,75 +3,64 @@ package mesosphere.marathon.state
 import com.wix.accord.dsl._
 import com.wix.accord._
 import mesosphere.marathon.api.v2.Validation._
-import org.apache.mesos.{ Protos => Mesos }
+
+import org.apache.mesos.Protos.ContainerInfo
+
 import scala.collection.immutable.Seq
 
-// TODO: trait Container and specializations?
-// Current implementation with type defaulting to DOCKER and docker to NONE makes no sense
-case class Container(
-    `type`: Mesos.ContainerInfo.Type = Mesos.ContainerInfo.Type.DOCKER,
-    volumes: Seq[Volume] = Nil,
-    docker: Option[Container.Docker] = None) {
+trait Container {
+  val volumes: Seq[Volume]
+
+  def forDocker(f: Container.Docker => Unit): Unit = {
+    this match {
+      case docker: Container.Docker => f(docker)
+      case _ =>
+    }
+  }
+
+  def foldDocker[R](defaultResult: R)(f: Container.Docker => R): R = {
+    this match {
+      case docker: Container.Docker => f(docker)
+      case _ => defaultResult
+    }
+  }
+
+  def mapDocker(f: Container.Docker => Container.Docker): Container = {
+    foldDocker(this)(f(_))
+  }
 
-  import Container._
+  def docker(): Option[Container.Docker] = foldDocker[Option[Container.Docker]](None)(Some(_))
 
-  def portMappings: Option[Seq[Docker.PortMapping]] = {
-    import Mesos.ContainerInfo.DockerInfo.Network
+  def dockerPortMappings: Option[Seq[Container.Docker.PortMapping]] = {
```

Comment:
Hm. That breaks a lot of tests, respectively if I change the field name then runtime error messages will look different. This could be avoided by refactoring a lot of code to use the parent's function instead of the field. I suggest to leave this for another PR. There are other cleanups that can be bundled with that then. For now, renaming the parent's function to getPortMappings. No conflicts with anything this way.